### ISSUE TITLE
TorchSharp models exhibit higher loss than identical PyTorch models during training

### ISSUE THREAD

**Issue by brianberns** [xamarin/TorchSharp#1082]
I'm working on an F# port of Andrej Karpathy's [minGPT](https://github.com/karpathy/minGPT) from PyTorch to TorchSharp. The port (called [MinGptSharp](https://github.com/brianberns/MinGptSharp])) is intended to match the original version as closely as possible, but I've noticed the resulting models perform worse than identical models built with minGPT. To be specific, the training loss for a TorchSharp/MinGptSharp model is significantly higher than the loss for the corresponding PyTorch/minGPT model after the same number of iterations. For example, for identicial character-level language models ([original](https://github.com/karpathy/minGPT/tree/master/projects/chargpt), [port](https://github.com/brianberns/MinGptSharp/tree/master/CharGpt)) trained on [Tiny Shakespeare](https://github.com/karpathy/char-rnn/blob/master/data/tinyshakespeare/input.txt), the PyTorch version has a training loss of 1.99580 after 500 iterations, while the corresponding TorchSharp model has a training loss of 2.42897.

It's entirely possible that I made mistakes in porting the code to F# and TorchSharp, but I've spent multiple hours stepping through both models side-by-side, and can find no discrepancies. To debug the problem, I switched to a simpler project that sorts numbers ([original](https://github.com/karpathy/minGPT/blob/master/demo.ipynb), [port](https://github.com/brianberns/MinGptSharp/tree/master/Demo)), and carefully tweaked both versions so they were totally deterministic: both had the same random seed, training data, and initial weights. After the initial forward pass, both models produced identical logits and loss, down to the last decimal digit. However, after a few iterations, the models began to diverge, with the PyTorch model always having a smaller loss. I suspect the divergence occurs either during back-propagation (`loss.backward()`) or optimizer descent (`optimizer.step()`), but I haven't been able to nail it down exactly.

Other possible differences I've considered include:

* **Package versions**. I'm using version 2.0.1 of PyTorch and version 0.100.4 of TorchSharp. It's hard to see how this could lead to the differences I'm seeing, though. The GPT models I'm running should behave essentially the same on any underlying version of LibTorch.
* **Optimizers**. I tried switching to a different optimizer, but the problem persisted.
* **Data loaders**. I configured both loaders to be deterministic in order to eliminate them as a possible difference. I also tried running both loaders with shuffling enabled, which produces better results, but the TorchSharp model is still worse.
* **CPU vs. CUDA**. I configured both versions to run on GPU only, in order to eliminate CUDA as a source of error.
* **Rounding errors**. Both versions use LibTorch on the same hardware and operating system (Windows 11), so should exhibit the same rounding behavior (as evidenced by the identical loss and logits of the first forward pass through the model).

I understand that this might not be a TorchSharp issue, per se, but I'm posting it here in case anyone has any ideas. For me, it's a showstopper at the moment, so I'd certainly appreciate guidance. I will continue investigating and will update this issue if I find anything. Thanks.

**Comment by lostmsu** [comment]
I had one for a while now and I believe it did not have major differences in loss. You can use it as a sanity check to validate if the issue is somehow with PyTorch or with your implementation: https://github.com/losttech/Torch.MinGPT

I noticed you have your own BPE implementation (mine does not, would love to see your BPE as a NuGet package). Could it be the source of the difference?

**Comment by brianberns** [comment]
Thank you, I will check that out.

My BPE is a port of the one in Karpathy's repos, but it turns out that none of his code actually uses it, so I haven't had a chance to use mine either. (The language model that has this problem uses character-level encoding, so BPE isn't needed.)

**Comment by brianberns** [comment]
Quick update for anyone who might be interested in this: I have confirmed that the gradients are different after the first backward pass, even though the loss and all the parameters are identical between the two models.

**Comment by brianberns** [comment]
Well, f*ck me, that is indeed the problem. Calling `model.zero_grad()` doesn't seem to work. I switched to `optimizer.zero_grad()` instead, and the outputs are not the same. I'm going to close this issue and open another one with a simple repro example.

**Comment by brianberns** [comment]
.
